src/Orleans.Runtime/Catalog/ActivationData.cs

```diff
@@ -391,6 +411,68 @@ private void ScheduleOperation(object operation)
             _workSignal.Signal();
         }
 
+        public void Migrate(Dictionary<string, object> requestContext, CancellationToken? cancellationToken = default)
+        {
+            if (!cancellationToken.HasValue)
+            {
+                cancellationToken = new CancellationTokenSource(_shared.InternalRuntime.CollectionOptions.Value.DeactivationTimeout).Token;
+            }
+
+            ScheduleOperation(new Command.Migrate(requestContext, cancellationToken.Value));
+        }
+
+        private async Task StartMigratingAsync(Dictionary<string, object> requestContext, CancellationToken cancellationToken)
+        {
+            // Run placement to select a new host. If a new (different) host is not selected, do not migrate.
+            var placementService = _shared.Runtime.ServiceProvider.GetRequiredService<PlacementService>();
+            var newLocation = await placementService.PlaceGrainAsync(GrainId, requestContext, PlacementStrategy);
+            if (newLocation == Address.SiloAddress || newLocation is null)
+            {
+                // No more appropriate silo was selected for this grain, but we cannot cancel migration now, so we must continue to deactivate and reactivate the grain.
```

Comment:
This is the first thing which happens when migrating, so the cost is just the cost of running placement, which is cheap.
I've updated the comment, since this one is now inaccurate: the migration will be canceled, and the grain will not be deactivated.

Reply:
I didn't answer your actual point: do we want to somehow prevent the grain code from spamming `MigrateOnIdle()`? Maybe. I'm not sure what the behavior should be. It seems like throwing wouldn't be useful. We could make the method async and return `ValueTask<bool>` or something else (with an implementation/design change). Something makes me uncomfortable about that. Another option is to ignore migration requests if they fall within some time period, but I think the added complexity is not worth it for the small benefit that you gain (not re-running placement, which is typically cheap).

Reply:
> I didn't answer your actual point: do we want to somehow prevent the grain code from spamming `MigrateOnIdle()`? 

Would it make sense to throw if there's only a single silo?